packages/devtools_shared/lib/src/utils/file_utils.dart

```diff
@@ -2,15 +2,78 @@
 // Use of this source code is governed by a BSD-style license that can be
 // found in the LICENSE file.
 
+import 'package:dtd/dtd.dart';
+import 'package:meta/meta.dart';
+import 'package:path/path.dart' as p;
+
 /// Attempts to detect the package root of [fileUriString].
 ///
-/// This method uses the heuristics that a Dart executable within a Dart package
-/// should reside in one of the following top-level folders: 'lib', 'bin',
-/// 'integration_test', 'test', or 'benchmark'.
-String packageRootFromFileUriString(String fileUriString) {
-  // TODO(kenz): for robustness, consider sending the root library uri to the
-  // server and having the server look for the package folder that contains the
-  // `.dart_tool` directory.
+/// This method first tries to use the Dart Tooling Daemon to walk the
+/// directory structure of [fileUriString] and look for the package root; we
+/// consider a directory to be a package root if it contains the `.dart_tool`
+/// directory.
+///
+/// If we cannot find the package root using the Dart Tooling Daemon, we use the
+/// heuristic that a Dart executable within a Dart package should reside in one
+/// of the following top-level folders:
+///
+/// * lib
+/// * bin
+/// * integration_test
+/// * test
+/// * benchmark
+/// * example
+///
+/// The URI returned will be a file URI String and will NOT have a trailing
+/// slash.
+Future<String> packageRootFromFileUriString(
+  String fileUriString, {
+  DartToolingDaemon? dtd,
+  @visibleForTesting bool throwOnDtdSearchFailed = false,
+}) async {
+  if (dtd != null) {
+    // Use type [Object] so we can store exceptions of all types.
+    Object? exception;
+
+    final parts = List.of(Uri.parse(fileUriString).pathSegments);
+    while (parts.isNotEmpty) {
+      parts.removeLast();
+      // Include the leading slash because `pathSegments` will remove it, and a
+      // leading slash is required for 'Uri.file' to properly set the 'file'
+      // URI scheme.
+      // Include the trailing slash because the IDE workspace root this URI is
+      // a part of may have a trailing slash, and DTD checks that the URI we
+      // pass to `listDirectoryContents` starts with at least one IDE workspace
+      // root.
+      var pathString = p.joinAll(parts);
+      if (!pathString.startsWith('/')) pathString = '/$pathString';
+      if (!pathString.endsWith('/')) pathString = '$pathString/';
+
+      final currentUri = Uri.file(pathString);
```

Comment:
This doesn't seem right for Windows, we shouldn't be calling `Uri.file()` with a string that contains `/` (or starts with) for Windows paths.

Since we're sending a URI to DTD anyway, I think we could just manipulate the URIs something like:

```dart
while (uri.pathSegments.length > 1) {
  // Take the last path segment off
  uri = uri.replace(pathSegments: uri.pathSegments.sublist(0, uri.pathSegments.length - 1));

  // uri is now the parent directory
}
```

That way we don't need to worry about which kind of paths they are (so if this code runs in the browser, we don't need to worry that it will assume non-Windows).

Reply:
Applied requested changes, PTAL.

Reply:
Thanks, yes if you wouldn't mind running all devtools_shared/test on windows with this change, that would be great, thank you!

Reply:
As-is I see a bunch of errors trying to clean delete previous test apps:

```
00:02 +38 -1: test/server\general_api_test.dart: General DevTools server API endpoints that require DTD detectRootPackageForVmService succeeds for a connect event [E]
  PathAccessException: Deletion failed, path = 'tmp/test_app' (OS Error: The process cannot access the file because it is being used by another process.
  , errno = 32)
  dart:io                                   FileSystemEntity.deleteSync
  test\helpers.dart 111:43                  TestDartApp.kill
  ===== asynchronous gap ===========================
  test\server\general_api_test.dart 176:11  main.<fn>.<fn>.<fn>.<fn>
```

This isn't uncommon on Windows - things are often locked for a while after a process dies and many of our tools retry these kinds of deletes.

I put a loop around the delete until it succeeded and everything passes:

```
PS C:\Dev\Google\devtools\packages\devtools_shared> ..\..\tool\flutter-sdk\bin\dart test test/
Resolving dependencies in `C:\Dev\Google\devtools\packages\devtools_shared`...
Downloading packages...
Got dependencies in `C:\Dev\Google\devtools\packages\devtools_shared`.
Building package executable...
Built test:test.
00:00 +22: test/server\deeplink_api_test.dart: DeeplinkApi handle deeplink api api/androidBuildVariants returns bad request if no root path
[api/androidBuildVariants] missing required query parameters: [rootPath]
00:00 +31: test/server\general_api_test.dart: General DevTools server API api/notifyForVmServiceConnection returns badRequest for invalid VM service argument
Cannot normalize VM service URI: fake_uri
00:00 +34: test/server\general_api_test.dart: General DevTools server API api/notifyForVmServiceConnection returns badRequest for invalid connected argument
Cannot parse connected parameter:
FormatException: Invalid boolean
bad_arg
00:02 +38: test/server\general_api_test.dart: General DevTools server API endpoints that require DTD detectRootPackageForVmService succeeds for a connect event
Failed to delete old test app on attempt 2... retrying...
Failed to delete old test app on attempt 3... retrying...
Failed to delete old test app on attempt 4... retrying...
Failed to delete old test app on attempt 5... retrying...
Failed to delete old test app on attempt 6... retrying...
00:08 +39: test/server\general_api_test.dart: General DevTools server API endpoints that require DTD detectRootPackageForVmService succeeds for a disconnect event when cache is empty
Failed to delete old test app on attempt 2... retrying...
00:10 +40: test/server\general_api_test.dart: General DevTools server API endpoints that require DTD detectRootPackageForVmService succeeds for a disconnect event when cache contains entry for VM service
Failed to delete old test app on attempt 2... retrying...
Failed to delete old test app on attempt 3... retrying...
Failed to delete old test app on attempt 4... retrying...
Failed to delete old test app on attempt 5... retrying...
Failed to delete old test app on attempt 6... retrying...
Failed to delete old test app on attempt 7... retrying...
00:17 +41: All tests passed!
```

I'll make a note to open a PR with a retry like this so it doesn't need to be done ad-hoc to run tests.

Reply:
I opened https://github.com/flutter/devtools/pull/7544 with a retry for the deletes so we only fail if we can't delete for some period (currently 10s).